### PR TITLE
Lifecycle rule enhancement for bucket level operation

### PR DESCRIPTION
**Pull request by PravinRanjan10**
This PR is to support lifecycle transition across multiple cloud vendors. Now, a user can specify target cloud with target buckets.
And the expection would be, after specific days object/file will be moved to target buckets of target backends. 
To experience this feature, please take a look of uploaded video below.

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind new feature
> /kind bug fix
> /kind cleanup
> /kind revert change
> /kind design
> /kind documentation
> /kind enhancement

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Test Report Added?**:
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
> /kind TESTED
> /kind NOT-TESTED

**Test Report**:

**Special notes for your reviewer**:

TestReports/Demo video:
https://www.youtube.com/watch?v=iregvviEenw

Inter + intra transition of lifecycle:
lifecycle-2

### CODE REVIEW

**Comment by kumarashit** [comment]
Please update PR details and add test report

**Comment by kumarashit** [comment]
Please update test report for different cloud providers

File: s3/pkg/datastore/azure/azure.go [hunk 1]
```diff
@@ -241,8 +241,9 @@ func (ad *AzureAdapter) Copy(ctx context.Context, stream io.Reader, target *pb.O
 
 func (ad *AzureAdapter) ChangeStorageClass(ctx context.Context, object *pb.Object, newClass *string) error {
 	objectId := object.ObjectId
-	blobURL := ad.containerURL.NewBlockBlobURL(objectId)
-	log.Infof("change storage class[Azure Blob], objectId:%s, blobURL is %v\n", objectId, blobURL)
+	containerURL, _ := ad.createBucketContainerURL(ad.backend.Access, ad.backend.Security, object.BucketName)
```
Comment (asifdxtreme) [inline s3/pkg/datastore/azure/azure.go:244]: Why do we skip the error checking here, createBucketContainerURL can return error too
```
func (ad *AzureAdapter) createBucketContainerURL(accountName string, accountKey string, bucketName string) (azblob.ContainerURL, error) 
```

File: s3/pkg/service/bucket.go [hunk 2]
```diff
@@ -84,18 +84,21 @@ func (s *s3Service) CreateBucket(ctx context.Context, in *pb.Bucket, out *pb.Bas
 		return err
 	}
 
-	sd, err := driver.CreateStorageDriver(backend.Type, backend)
-	if err != nil {
-		log.Errorln("failed to create storage. err:", err)
-		return err
-	}
+	if backend.Type == "aws-s3" || backend.Type == "azure-blob" || backend.Type == "gcp-s3" ||
```
Comment (asifdxtreme) [inline s3/pkg/service/bucket.go:87]: Can we improve on this condition, something like backend is cloud Type then condition is true

**Review by asifdxtreme — APPROVED** [review]
LGTM

File: s3/pkg/datastore/azure/azure.go (hunk 1 above)
Comment (PravinRanjan10) [reply s3/pkg/datastore/azure/azure.go]: done. Thanks

File: s3/pkg/service/bucket.go (hunk 2 above)
Comment (PravinRanjan10) [reply s3/pkg/service/bucket.go]: done. Thanks

**Comment by rajat-soda** [comment]
LGTM

File: s3/pkg/service/bucket.go [hunk 3]
```diff
@@ -30,6 +30,17 @@ import (
 var ENC_KEY_LEN int = 32
 var ENC_IV_LEN int = 16
 
+var CrudSupportedClouds = []string{"aws-s3", "azure-blob", "gcp-s3", "hw-obs"}
```
[the remark this answers is not in the document]
Comment (AmitRoushan) [reply s3/pkg/service/bucket.go:33]: Also the datastructure can be optimized to have map[string]struct{}. will help in contain function

File: s3/pkg/service/bucket.go [hunk 4]
```diff
@@ -30,6 +30,17 @@ import (
...
+
+func contains(slice []string, item string) bool {
```
Comment (AmitRoushan) [inline s3/pkg/service/bucket.go:35]: function name can be optimized as ```isCRUCSupportedCloud``` as contain function is used for the purpose only

File: s3/pkg/service/bucket.go [hunk 5]
```diff
@@ -241,18 +254,20 @@ func (s *s3Service) DeleteBucket(ctx context.Context, in *pb.Bucket, out *pb.Bas
...
-	}
+    if contains(CrudSupportedClouds, backend.Type){
+		sd, err := driver.CreateStorageDriver(backend.Type, backend)
+		if err != nil {
+			log.Errorln("failed to create storage. err:", err)
+			return err
+		}
 
-	log.Info("The driver for the backend is:\n", sd)
+		log.Info("The driver for the backend is:\n", sd)
```
Comment (AmitRoushan) [inline s3/pkg/service/bucket.go:264]: ```\n``` is redundant here

File: s3/pkg/service/object.go [hunk 6]
```diff
@@ -817,12 +822,14 @@ func (s *s3Service) MoveObject(ctx context.Context, in *pb.MoveObjectRequest, ou
...
 	} else {
+		log.Infof("Now need to move data and also change storage tier:%s", targetObject.ObjectKey)
 		// need move data, get target location first
 		if in.MoveType == utils.MoveType_ChangeLocation {
 			targetBucket = srcBucket
 			targetObject.Location = in.TargetLocation
-			log.Infof("move %s cross backends, srcBackend=%s, targetBackend=%s, targetTier=%d\n",
-				srcObject.ObjectKey, srcObject.Location, targetObject.Location, targetObject.Tier)
+			targetObject.BucketName = in.TargetBucket
+			log.Infof("move %s cross backends, srcBackend=%s, targetBackend=%s, targetBucket=%s targetTier=%d\n",
```
Comment (AmitRoushan) [inline s3/pkg/service/object.go:831]: Please follow info log convention as mentioned above

**Review by AmitRoushan — CHANGES_REQUESTED** [review]

File: s3/pkg/datastore/azure/azure.go [hunk 7]
```diff
@@ -241,8 +241,13 @@ func (ad *AzureAdapter) Copy(ctx context.Context, stream io.Reader, target *pb.O
...
 	objectId := object.ObjectId
-	blobURL := ad.containerURL.NewBlockBlobURL(objectId)
-	log.Infof("change storage class[Azure Blob], objectId:%s, blobURL is %v\n", objectId, blobURL)
+	containerURL, urlerr := ad.createBucketContainerURL(ad.backend.Access, ad.backend.Security, object.BucketName)
+	if urlerr != nil {
+		log.Errorf("failed to create bucket container url for azure:%v\n", urlerr)
+		return urlerr
+	}
+	blobURL := containerURL.NewBlobURL(objectId)
+	log.Infof("change storage class[Azure Blob], object=[%s], objectId:%s, blobURL is %v\n", object, objectId, blobURL)
```
[the remark this answers is not in the document]
Comment (PravinRanjan10) [reply s3/pkg/datastore/azure/azure.go]: done. Thanks

File: s3/pkg/service/bucket.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (PravinRanjan10) [reply s3/pkg/service/bucket.go]: done. Thanks

File: s3/pkg/service/bucket.go (hunk 5 above)
Comment (PravinRanjan10) [reply s3/pkg/service/bucket.go]: done. Thanks

File: s3/pkg/service/object.go (hunk 6 above)
Comment (PravinRanjan10) [reply s3/pkg/service/object.go]: done. Thanks

File: s3/pkg/service/bucket.go [hunk 8]
```diff
@@ -30,6 +30,17 @@ import (
 var ENC_KEY_LEN int = 32
 var ENC_IV_LEN int = 16
 
+var CRUDSupportedClouds = []string{"aws-s3", "azure-blob", "gcp-s3", "hw-obs"}
+
+func isCrudSupportedCloud(slice []string, item string) bool {
```
Comment (AmitRoushan) [inline s3/pkg/service/bucket.go:35]: just my opinion:
better implementation might be :
```
var CRUDSupportedClouds = map[string]struct{}{
                                   "aws-s3": struct{}{},
                                    "azure-blob": struct{}{},
                                    "gcp-s3": struct{}{}, 
                                    "hw-obs": struct{}{},
                               }
func isCrudSupportedCloud(cloud string) bool {
     _, ok := CRUDSupportedClouds[cloud]
     return ok
}

```

File: dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go [hunk 9]
```diff
@@ -329,8 +329,8 @@ func schedSortedActionsRules(inRules *InterRules) {
 						continue
 					}
 					log.Infof("lifecycle action: object=[%s] type=[%d] source-tier=[%d] target-tier=[%d] "+
-						"source-backend=[%s] target-backend=[%s].\n", obj.ObjectKey, r.ActionType, obj.Tier, r.Tier,
-						obj.Location, r.Backend)
+						"source-backend=[%s] target-backend=[%s] and targetBucket=[%s].\n", obj.ObjectKey, r.ActionType, obj.Tier, r.Tier,
```
Comment (AmitRoushan) [inline dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go:332]: IMO, lot of places redundant ```\n``` is used. Please clean them

**Review by AmitRoushan — CHANGES_REQUESTED** [review]

File: s3/pkg/service/bucket.go (hunk 8 above)
Comment (asifdxtreme) [reply s3/pkg/service/bucket.go:35]: +1

**Review by kumarashit — APPROVED** [review]
LGTM

**Pull request merged**